src/java.base/share/classes/java/lang/runtime/SwitchBootstraps.java

```diff
@@ -389,41 +315,210 @@ private static <T extends Enum<T>> int mappedEnumLookup(T value, MethodHandles.L
         return enumMap.map[value.ordinal()];
     }
 
-    private static boolean enumEqCheck(Object value, EnumDesc<?> label, MethodHandles.Lookup lookup, ResolvedEnumLabel resolvedEnum) {
-        if (resolvedEnum.resolvedEnum == null) {
-            Object resolved;
+    private static MethodHandle withIndexCheck(MethodHandle target, int labelsCount) {
+        MethodHandle checkIndex = MethodHandles.insertArguments(CHECK_INDEX, 1, labelsCount + 1);
 
-            try {
-                Class<?> clazz = label.constantType().resolveConstantDesc(lookup);
+        return MethodHandles.filterArguments(target, 1, checkIndex);
+    }
 
-                if (value.getClass() != clazz) {
-                    return false;
-                }
+    private static final class ResolvedEnumLabels implements BiFunction<Integer, Object, Object> {
 
-                resolved = label.resolveConstantDesc(lookup);
-            } catch (IllegalArgumentException | ReflectiveOperationException ex) {
-                resolved = SENTINEL;
-            }
+        private final MethodHandles.Lookup lookup;
+        private final EnumDesc<?>[] enumDescs;
+        @Stable
+        private Object[] resolvedEnum;
 
-            resolvedEnum.resolvedEnum = resolved;
+        public ResolvedEnumLabels(MethodHandles.Lookup lookup, EnumDesc<?>[] enumDescs) {
+            this.lookup = lookup;
+            this.enumDescs = enumDescs;
+            this.resolvedEnum = new Object[enumDescs.length];
         }
 
-        return value == resolvedEnum.resolvedEnum;
-    }
+        @Override
+        public Object apply(Integer labelIndex, Object value) {
+            Object result = resolvedEnum[labelIndex];
 
-    private static MethodHandle withIndexCheck(MethodHandle target, int labelsCount) {
-        MethodHandle checkIndex = MethodHandles.insertArguments(CHECK_INDEX, 1, labelsCount + 1);
+            if (result == null) {
+                try {
+                    EnumDesc<?> label = enumDescs[labelIndex];
+                    Class<?> clazz = label.constantType().resolveConstantDesc(lookup);
 
-        return MethodHandles.filterArguments(target, 1, checkIndex);
-    }
+                    if (value.getClass() != clazz) {
```

Comment:
Not related to this patch, but this appears to be wrong: we should do something like
```java
if (!(value instanceof Enum<?> ev) || ev.getDeclaringClass() != clazz)
    return SENTINEL; // or false
```

Reply:
You are right, of course, but let's solve that under [JDK-8318144](https://bugs.openjdk.org/browse/JDK-8318144), to ensure that can more easily be backported. Tests are running on the fix, will publish PR once they pass. Thanks!

Reply:
FYI:
https://github.com/openjdk/jdk/pull/16499